File: playwright-csharp/tests/Steps/RestfulBooker/API/BookingApiSteps.cs
using System.Text.Json.Nodes;
using NUnit.Framework;
using Reqnroll;
using TestTracks.Playwright.CSharp.Specs.Data;
using TestTracks.Playwright.CSharp.Specs.Data.DataBuilders.RestfulBooker;
using TestTracks.Playwright.CSharp.Specs.Support;
using AuthApiClient = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Auth.AuthApi;
using AuthToken = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Auth.AuthToken;
using BookingApiClient = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Booking.BookingApi;
using CreatedBooking = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Booking.CreatedBooking;
using RoomApiClient = TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Room.RoomApi;

namespace TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API;

[Binding]
public sealed class BookingApiSteps
{
    private const string ApiTarget = "restfulBookerApi";
    private const string AdminCredential = "restfulBookerAdmin";

    private readonly ScenarioState _state;
    private AuthToken? _authToken;
    private JsonObject? _bookingPayload;
    private CreatedBooking? _createdBooking;

    public BookingApiSteps(ScenarioState state)
    {
        _state = state;
    }

    [Given("valid booking details")]
    public async Task GivenValidBookingDetails()
    {
        var data = _state.Data.Load<BookingApiTestData>(
            "api-booking",
            "scenarios.validBookingDetails");

        var authApi = await _state.UseApiAsync(
            ApiTarget,
            (settings, request, evidence) => new AuthApiClient(settings, request, evidence));
        var roomApi = await _state.UseApiAsync(
            ApiTarget,
            (settings, request, evidence) => new RoomApiClient(settings, request, evidence));

        var admin = _state.Targets.Credential(AdminCredential);

        _authToken = await authApi.LogIn(AuthPayloadBuilder.FromCredential(admin));

        var room = RoomDataBuilder
            .ForScenario(_state.ScenarioId, data.Room)
            .BuildPayload();
        var createdRoom = await roomApi.CreateRoom(room, _authToken.Token);
        _state.Cleanup.Register(
            $"Delete room {createdRoom.RoomId}",
            () => roomApi.DeleteRoom(createdRoom.RoomId, _authToken!.Token));

        _bookingPayload = BookingDataBuilder
            .ForScenario(_state.ScenarioId, data.Booking)
            .ForRoom(createdRoom.RoomId)
            .BuildPayload();
    }

    [When("the booking is created through the booking API")]
    public async Task WhenTheBookingIsCreatedThroughTheBookingApi()
    {
        var bookingApi = await _state.UseApiAsync(
            ApiTarget,
            (settings, request, evidence) => new BookingApiClient(settings, request, evidence));

        _createdBooking = await bookingApi.CreateBooking(
            _bookingPayload!,
            _authToken!.Token);

        _state.Cleanup.Register(
            $"Delete booking {_createdBooking.BookingId}",
            () => bookingApi.DeleteBooking(_createdBooking!.BookingId, _authToken!.Token));
    }

    [Then("the booking can be retrieved with the same details")]
    public async Task ThenTheBookingCanBeRetrievedWithTheSameDetails()
    {
        var bookingPayload = _bookingPayload
            ?? throw new InvalidOperationException("Booking payload was not created.");

        var bookingApi = await _state.UseApiAsync(
            ApiTarget,
            (settings, request, evidence) => new BookingApiClient(settings, request, evidence));

        var retrieved = await bookingApi.GetBooking(
            _createdBooking!.BookingId,
            _authToken!.Token);

        Assert.That(retrieved.FirstName, Is.EqualTo(bookingPayload.StringValue("firstname")));
        Assert.That(retrieved.LastName, Is.EqualTo(bookingPayload.StringValue("lastname")));
        Assert.That(retrieved.RoomId, Is.EqualTo(bookingPayload.IntValue("roomid")));
    }
}

public sealed record BookingApiTestData(
    BookingDataDefaults Booking,
    RoomDataDefaults Room);

File: playwright-java/src/test/java/testtracks/steps/restfulbooker/api/BookingApiSteps.java
package testtracks.steps.restfulbooker.api;

import com.fasterxml.jackson.databind.node.ObjectNode;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.junit.jupiter.api.Assertions;
import testtracks.data.JsonPayloads;
import testtracks.data.builders.restfulbooker.AuthPayloadBuilder;
import testtracks.data.builders.restfulbooker.BookingDataBuilder;
import testtracks.data.builders.restfulbooker.BookingDataDefaults;
import testtracks.data.builders.restfulbooker.RoomDataBuilder;
import testtracks.data.builders.restfulbooker.RoomDataDefaults;
import testtracks.steps.restfulbooker.api.auth.AuthApi;
import testtracks.steps.restfulbooker.api.auth.AuthToken;
import testtracks.steps.restfulbooker.api.booking.BookingApi;
import testtracks.steps.restfulbooker.api.booking.CreatedBooking;
import testtracks.steps.restfulbooker.api.room.RoomApi;
import testtracks.support.ScenarioState;

public final class BookingApiSteps {
  private static final String API_TARGET = "restfulBookerApi";
  private static final String ADMIN_CREDENTIAL = "restfulBookerAdmin";

  private final ScenarioState state;
  private AuthToken authToken;
  private ObjectNode bookingPayload;
  private CreatedBooking createdBooking;

  public BookingApiSteps(ScenarioState state) {
    this.state = state;
  }

  @Given("valid booking details")
  public void validBookingDetails() {
    var data =
        state.data().load("api-booking", "scenarios.validBookingDetails", BookingApiTestData.class);
    var authApi = state.useApi(API_TARGET, AuthApi::new);
    var roomApi = state.useApi(API_TARGET, RoomApi::new);
    var admin = state.targets().credential(ADMIN_CREDENTIAL);

    authToken = authApi.logIn(AuthPayloadBuilder.fromCredential(admin));

    var room = RoomDataBuilder.forScenario(state.scenarioId(), data.room()).buildPayload();
    var createdRoom = roomApi.createRoom(room, authToken.token());
    state
        .cleanup()
        .register(
            "Delete room " + createdRoom.roomId(),
            () -> roomApi.deleteRoom(createdRoom.roomId(), authToken.token()));

    bookingPayload =
        BookingDataBuilder.forScenario(state.scenarioId(), data.booking())
            .forRoom(createdRoom.roomId())
            .buildPayload();
  }

  @When("the booking is created through the booking API")
  public void theBookingIsCreatedThroughTheBookingApi() {
    var bookingApi = state.useApi(API_TARGET, BookingApi::new);
    createdBooking = bookingApi.createBooking(bookingPayload, authToken.token());

    state
        .cleanup()
        .register(
            "Delete booking " + createdBooking.bookingId(),
            () -> bookingApi.deleteBooking(createdBooking.bookingId(), authToken.token()));
  }

  @Then("the booking can be retrieved with the same details")
  public void theBookingCanBeRetrievedWithTheSameDetails() {
    var bookingApi = state.useApi(API_TARGET, BookingApi::new);
    var retrieved = bookingApi.getBooking(createdBooking.bookingId(), authToken.token());

    Assertions.assertEquals(
        JsonPayloads.stringValue(bookingPayload, "firstname"), retrieved.firstName());
    Assertions.assertEquals(
        JsonPayloads.stringValue(bookingPayload, "lastname"), retrieved.lastName());
    Assertions.assertEquals(JsonPayloads.intValue(bookingPayload, "roomid"), retrieved.roomId());
  }

  public record BookingApiTestData(BookingDataDefaults booking, RoomDataDefaults room) {}
}
